### DISCUSSION TITLE
Concurrency: Number of CPUs?

### DISCUSSION THREAD

**Discussion by wilhelmer** [squidfunk/mkdocs-material#4998] [Q&A] [answered]
From https://squidfunk.github.io/mkdocs-material/setup/building-an-optimized-site/#built-in-optimize-plugin:

> [concurrency](https://squidfunk.github.io/mkdocs-material/setup/building-an-optimized-site/#+optimize.concurrency)
> Default: number of CPUs – This option specifies how many CPUs the plugin is allowed to use when optimizing media files.

What does "Default: number of CPUs" mean? Does that mean that Material detects how many CPU cores are in my system and sets the option to that number?

If yes, I think it would be better to document this as, e.g., `Default: number of CPUs detected`.

**Comment by squidfunk** [comment, marked as answer]
> What does "Default: number of CPUs" mean? Does that mean that Material detects how many CPU cores are in my system and sets the option to that number?

Yes, it's equal to the number of CPUs on the build system. Detection is done using `os.cpu_count()`. I'm not happy with `Default: number of CPUs detected` either, I'll see if we can find a better formulation.

**Reply by wilhelmer** [reply]
Maybe "number of CPUs in the system", following the [official documentation for cpu_count()](https://docs.python.org/3/library/os.html#os.cpu_count).

**Comment by wilhelmer** [comment]
Okay, thanks for the explanation. So in most cases, it's recommended to keep the default setting.

**Reply by squidfunk** [reply]
Yes, definitely. The optimize plugin and privacy plugin have intelligent caching in-place, so repetitive invocations are cheap 😊 Thus, using more CPUs will make the first build much faster. Note that the privacy plugin was just updated to also support concurrency, and caching was significantly improved, now also providing support for alternate query strings in URLs.
